### DISCUSSION TITLE
session flash data is missing when using redirect with nested component and Modelable

### DISCUSSION THREAD

**Discussion by jhonoryza** [livewire/livewire#7797] [7. Bugs (archived)] [answered]
### Livewire version

3.4.1

### Laravel version

10.42.0

### Which PHP version are you using?

PHP 8.2

### Steps To Reproduce

i found a weird behaviour when using session flash with redirect in nested component with Modelable

let say i have a livewire component `SettingForm`

after form is submitted we call save function like this

```php
public function save() {
        session()->flash('alertType', 'success');
        session()->flash('alertMessage', 'setting is updated');

        return redirect()->to(route('cms.settings.index'));
}

public function render()
    {
        return view('livewire.setting_form')
            ->section('content');
    }
```
setting_form.blade.php
```php
<div>
                   <input type="text" wire:model="title" />
                    <livewire:file-upload wire:model="images" />
</div>
```

FileUpload component
```php
class FileUpload extends Component
{
    use WithFileUploads;

    #[Modelable]
    public $file;

    public function render()
    {
        return view('livewire.file-upload');
    }
}
```

after the page redirected 
if i comment the livewire component `<livewire:file-upload-component` everything is works as expected
but if i include`<livewire:file-upload-component` the session data is missing 
if i  comment / disable #[Modelable] everything is works as expected

i check in this class `Livewire\Features\SupportRedirects\SupportRedirects::class` have dehydrate function that clear the session flash data when there is no redirect

```php
      if (! $to) {
            // If there was no redirect. Clear flash session data.
            if (app()->has('session.store')) {
                session()->forget(session()->get('_flash.new'));
            }

            return;
        }
```

the problem is if i have a nested component with Modelable after redirect function is called, the dehydrate function for the nested component is called and `$to` variable is null so the session data is removed

can you suggest what should i do ? or am i missing something here ?

### Repository or Wirebox URL

https://github.com/jhonoryza/livewire-bug-redirect

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by perisicnikola37** [comment]
Hi there,

After analyzing your provided code and description, it seems like there might be a conflict in the dehydrate function within the Livewire\Features\SupportRedirects\SupportRedirects::class class. The issue arises when a nested component with Modelable is present, causing the $to variable to be null in the dehydrate function and subsequently leading to the removal of session flash data.

Here are a couple of suggestions to address this issue:

Update Livewire:
Ensure that you are using the latest version of Livewire. There might be bug fixes or improvements in newer releases that address the specific problem you are encountering.

Temporary Workaround:
As a temporary workaround, you might consider restructuring your Livewire components or the way you handle redirection. You could try using the with method for redirect and explicitly pass the session flash data, like so:

```
return redirect()->to(route('cms.settings.index'))->with([
    'alertType' => 'success',
    'alertMessage' => 'setting is updated',
]);
```
This way, you might bypass the session flash data removal triggered by the dehydrate function.

**Reply by jhonoryza** [reply]
@perisicnikola37 i already use livewire v3.4.2 latest version, i try what you suggest
![image](https://github.com/livewire/livewire/assets/5910636/6d25356e-7aea-4edd-9414-496a8f21ae0e)
![image](https://github.com/livewire/livewire/assets/5910636/1b5a0517-e7b9-437a-b6ee-e8b62e47adc7) 
![image](https://github.com/livewire/livewire/assets/5910636/d5c9672c-c66b-4e9c-a794-f30e502124a5)


still the session data is missing or removed

**Reply by perisicnikola37** [reply]
@jhonoryza So you are not seeing printing of $value in `detail.blade.php` file?

**Comment by jhonoryza** [comment]
> @jhonoryza So you are not seeing printing of $value in `detail.blade.php` file?

true i am not seeing it printing $value here

you can try the wirebox link above

**Comment by perisicnikola37** [comment]
The problem was in the file `"SettingForm.php"` `ln. 18` :)

You have to use this code to set a session with a certain name:
 `session()->put('alertMessage', 'Your custom alert message..');`
 I opened a pull request with fixed solution in your original project repository(https://github.com/jhonoryza/livewire-bug-redirect).. just merge it and everything will be fine :)

Please mark this answer as the **correct** response so we can close this discussion.

**Reply by jhonoryza** [reply]
@perisicnikola37  i dont think that is the solution, because `session()->flash()` and `session->put()` is a different thing :)

**Reply by perisicnikola37** [reply]
@jhonoryza 

Forget flashing via session when using Livewire.

It will only be stored, then retrieved from session on a full page request.

You need to use **events** for that.

**Reply by jhonoryza** [reply, marked as answer]
fixed in https://github.com/livewire/livewire/pull/7837
